From 1fbd95517884aa8902473faa36720eb0ab9b7576 Mon Sep 17 00:00:00 2001
From: Kouhei Ueno <kouhei@chromium.org>
Date: Tue, 28 May 2024 06:56:29 +0000
Subject: [PATCH] GetAcceptLanguages: Prospective crash fix

Querying `navigator.language` property on a destroyed context of a
worker can result in a nullptr deref.

This CL is a prospective fix to the problem.

Bug: 40945292, 40827704
Change-Id: If867f3e7cf188d48ff286d4228fc318ab723220b
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5568777
Auto-Submit: Kouhei Ueno <kouhei@chromium.org>
Reviewed-by: Hiroki Nakagawa <nhiroki@chromium.org>
Commit-Queue: Hiroki Nakagawa <nhiroki@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1306618}
---

--- a/third_party/blink/renderer/core/workers/worker_navigator.cc
+++ b/third_party/blink/renderer/core/workers/worker_navigator.cc
@@ -42,6 +42,13 @@
 
 String WorkerNavigator::GetAcceptLanguages() {
   auto* global_scope = To<WorkerOrWorkletGlobalScope>(GetExecutionContext());
+  if (!global_scope) {
+    // Prospective fix for crbug.com/40945292 and crbug.com/40827704
+    // Return empty string since it is better than crashing here, and the return
+    // value is not that important since the worker context is already
+    // destroyed.
+    return "";
+  }
 
   return global_scope->GetAcceptLanguages();
 }
